test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeTest.java

```diff
@@ -0,0 +1,344 @@
+            }
+
+            if (connectTimeout != null) {
+                env.put("com.sun.jndi.ldap.connect.timeout", connectTimeout.toString());
+            }
+            env.put(Context.SECURITY_PROTOCOL, "ssl");
+            env.put(Context.SECURITY_AUTHENTICATION, "Simple");
+            env.put(Context.SECURITY_PRINCIPAL, "cn=principal");
+            env.put(Context.SECURITY_CREDENTIALS, "123456");
+            LdapContext ctx = null;
+            try {
+                ctx = new InitialLdapContext(env, null);
+            } catch (NamingException e) {
+                if (customSocketFactory != null) {
+                    System.out.println("Caught expected Exception with custom SocketFactory (no SSL Socket).");
+                    if (CustomSocketFactory.customSocket.closeMethodCalledCount() <= 0) {
+                        System.err.println("Custom Socket was not closed.");
+                        System.exit(-1);
+                    }
+                } else if (connectTimeout > 0) {
+                    System.out.println("Caught expected Exception with connectTimeout > 0.");
+                } else {
+                    throw e;
+                }
+            } finally {
+                if (ctx != null) {
+                    System.out.println("Context was created, closing it.");
+                    Socket sock = getSocket(ctx);
+                    ctx.close();
+                    if (!checkSocketClosed(sock)) {
+                        System.exit(-1);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.exit(-1);
+        }
+    }
+
+    private static Socket getSocket(LdapContext ctx) throws Exception {
+        Field defaultInitCtxField = ctx.getClass().getSuperclass().getSuperclass().getDeclaredField("defaultInitCtx");
+        defaultInitCtxField.setAccessible(true);
+        Object defaultInitCtx = defaultInitCtxField.get(ctx);
+        Field clntField = defaultInitCtx.getClass().getDeclaredField("clnt");
+        clntField.setAccessible(true);
+        Object clnt = clntField.get(defaultInitCtx);
+        Field connField = clnt.getClass().getDeclaredField("conn");
+        connField.setAccessible(true);
+        Object conn = connField.get(clnt);
+        return (Socket)conn.getClass().getDeclaredField("sock").get(conn);
+    }
+
+    private static boolean checkSocketClosed(Socket s) throws Exception {
+        if (s.isClosed()) {
+            System.out.println("Socket is closed");
+            return true;
+        } else {
+            System.err.println("Socket isn't closed");
+            return false;
+        }
+    }
+
+    private static class CustomSocket extends Socket {
+        private int closeMethodCalled;
+
+        public CustomSocket() {
+            super();
+        }
+
+        public CustomSocket(String s, int timeout) throws IOException {
+            super(s, timeout);
```

Comment:
The argument should be called "port" not timeout.

Reply:
Correct.